DOC: Update einsum docs

attack68 opened the pull request:
I've noticed a few other issues and anecdotal reports of einsum being somewhat hard to grasp. I remember a while ago having those same experiences although now feel it is one of the most useful functions that I use consistently.

I wanted to expand the docs to make the function more understandable and likeable by first time readers, albeit I tried not to augment the current material too much so as to not step on anyones toes.

Also I added a part and example on where the einsum_path has been very useful for some of my optimisation routines and have increased speed by orders of magnitude.

attack68 committed:
Einsum docs improvements.

attack68 committed:
DOC: add information to numpy.core.einsumfunc

eric-wieser commented on numpy/core/einsumfunc.py:
```diff
@@ -884,39 +886,68 @@ def einsum(*operands, **kwargs):
+        * Matrix multiplication and vector dot product.
+        * Broadcasting and scalar multiplication.
+        * Tensor contractions.
+        * Chained array operations, in efficient calculation order.
```
"array operations" is pretty vague.

Also, does this list render correctly in sphinx?

attack68 replied:
Thanks for the overall feedback, Eric.

I agree it is vague, to be consistent with other paragraphs and for clarity perhaps "multi-dimensional, linear algebraic array operations" is better.

Sphinx rendering of bullet points is OK. I have copied the format from other numpy docstrings and tested against an online sphinx renderer.

mattip replied:
It will render without a vertical bar (quote mark) if you untab four spaces, aligning the `*` with the previous text. Also IMO `examples:` not `examples,`

attack68 replied:
adopted

eric-wieser commented on numpy/core/einsumfunc.py:
```diff
@@ -884,39 +886,68 @@ def einsum(*operands, **kwargs):
-    The order of labels in the output is by default alphabetical.  This
+    with no changes. A further example ``np.einsum('ij,jk', a, b)``
+    describes traditional matrix
+    multiplication and is equivalent to ``np.matmul(a,b)``. Repeated
```
This is true only when a and b are 2d.

attack68 replied:
Agreed but they are 2d by implication from the subscripts. Do you not consider it clearer for the reader to naturally assume 2d arrays in combination with matrix multiplication, rather than specifying the shape with additions to the description?
If your comment specifically regards `matmul` being capable of operating on arrays of other dimensions I would prefer to consider this out of scope for this function's docstring, since this example's purpose is only to demonstrate the familiar 'ij,jk' string format relative to matrix multiplication.

Alternatively I can just remove "and is equivalent to np.matmul(a,b)" but I do consider it to be of some value here?

eric-wieser commented on numpy/core/einsumfunc.py:
```diff
@@ -884,39 +886,68 @@ def einsum(*operands, **kwargs):
+    ``np.einsum('ij,jh', a, b)`` returns the transpose of the
+    multiplication since subscript 'h' precedes subscript 'i'.
+
+    The output can be directly controlled by invoking the explict mode
```
typo: explicit

attack68 replied:
fixed

eric-wieser commented on numpy/core/einsumfunc.py:
```diff
@@ -884,39 +886,68 @@ def einsum(*operands, **kwargs):
 
+    The Einstein summation convention can be used to compute
+    many multi-dimensional array operations. Additionally, the explicit
+    mode of this function provides further flexibility that can be used
```
It might be less confusing to clearly separate the explicit and non-explicit modes in the description

attack68 replied:
Since this was a new observation in the docs I tried to thread it neatly into the existing material. I think you are right though and maybe highlighting different modes needs to be more prominent.

eric-wieser commented on numpy/core/einsumfunc.py:
```diff
@@ -1039,6 +1105,8 @@ def einsum(*operands, **kwargs):
            [ 4796.,  5162.],
            [ 4928.,  5306.]])
 
+    Example in ellipsis use:
```
"Example of"

attack68 replied:
fixed

eric-wieser commented on numpy/core/einsumfunc.py:
```diff
@@ -1051,6 +1119,8 @@ def einsum(*operands, **kwargs):
 
+    Writeable returned arrays:
+
     >>> # since version 1.10.0
```
May as well move this comment to the line above

attack68 replied:
done

eric-wieser commented on numpy/core/einsumfunc.py:
```diff
@@ -1059,6 +1129,17 @@ def einsum(*operands, **kwargs):
+    einsum path insertion:
+
+    >>> # since version 1.12.0
+    >>> path, desc = np.einsum_path('ij,ki,lj->lk', a, b, a, optimize='optimal')
```
This line gives an error for me

attack68 replied:
fixed - required reordering

attack68 commented on numpy/core/einsumfunc.py:
```diff
@@ -1059,6 +1129,17 @@ def einsum(*operands, **kwargs):
+
+    >>> # since version 1.12.0
+    >>> path, desc = np.einsum_path('ij,ki,lj->lk', a, b, a, optimize='optimal')
+    >>> # potential use inside loop or iteration
```
This was my preferred option actually but felt it was too contrived, but happy to revert.

attack68 committed:
DOC: add information to numpy.core.einsumfunc

mattip commented on numpy/core/einsumfunc.py:
```diff
@@ -884,43 +893,73 @@ def einsum(*operands, **kwargs):
+    The Einstein summation convention can be used to compute
+    many multi-dimensional, linear algebraic array operations. `einsum`
+    provides a succinct way of representing these in a
+    clear and consistent format.
```
"in a clear and consistent format" is a bit judgemental and not needed. Could also be read as "if you don't grok the syntax it is your fault"

attack68 replied:
The general point I was trying to convey is that for algorithms that contain lots of different linear algebraic operations it can be clearer where this format is repetitively used.
But I also maintain the view that less verbose descriptions are generally better so I think you're right!

attack68 committed:
DOC: add information to numpy.core.einsumfunc

attack68 committed:
DOC: add information to numpy.core.einsumfunc

mattip commented:
Thanks for the improvements.
As long as we are cleaning this up, can I request two things?
We can create cross references, for instance `np.trace(a)` can link to the documentation of `trace` by using
```
:py:func:`np.trace(a) <numpy.trace>`
```
which preserves the arguments while producing an internal link. (Just for reference, if there are no args you could just write
```
:py:func:`numpy.trace`
```
)
It would be nice to do this for the 5 internal functions referenced: `np.trace`, `np.inner`, `np.sum`, `np.diag`, and `np.swapaxes`

Also, in `numpy.add_newdocs.py` there is documentation to `c_einsum` which not only duplicates this documentation, it is never used since `c_einsum` is no longer exported from `np.core`. Could you delete that whole section?

attack68 committed:
DOC: remove 'numpy.core','c_einsum' section from 'add_newdocs.py'

attack68 committed:
DOC: make :py:func links to all internal functions.

attack68 commented:
There were a few more internal functions referenced, and some in the examples which I collectively posted in the bulleted list. I hope the format/markdown is correct.

I also removed the redundant section in the add_newdocs.py file.

attack68 committed:
DOC: make :py:func links to all internal functions (minor corrections)

attack68 commented:
Does that mean replace the section in `numpy.add_newdocs.py` I was asked to remove, but ensure the same documentation material is put in place?

charris commented:
I think it means keep the section, but change the path from `numpy.core` to `numpy.core.multiarray`. Let me try and see what happens ...

charris commented:
Yep, that fixes the missing documentation.

attack68 committed:
DOC: place `add_newdoc('numpy.core.multiarray','c_einsum'`,...) in file.

mhvk commented on numpy/add_newdocs.py:
```diff
@@ -2265,28 +2265,89 @@ def luf(lamdaexpr, *args, **kwargs):
     operands : list of array_like
         These are the arrays for the operation.
-    out : ndarray, optional
+    out : {ndarray, None}, optional
```
I would leave this as it was - `None` is not really something one should pass in.

attack68 replied:
Actually this wasn't my change. You have observed a difference in the underlying docs (as a result of having the code in two places). When I copied over from `einsumfunc.py` it conflicts.
Anyway I agree and have now set them both to ndarray only.

mhvk commented on numpy/add_newdocs.py:
```diff
@@ -2319,101 +2385,188 @@ def luf(lamdaexpr, *args, **kwargs):
+    returns the optimal path in the majority of cases. In some cases 'optimal'
+    will return the superlative path through a more expensive, exhaustive search.
+    For iterative calculations it may be advisable to calculate the optimal path
+    once and reuse that path by supplying it as an argument. An example is given.
```
"is given *below*"

attack68 replied:
changed.

mattip replied:
This change seems missing?

mhvk commented on numpy/add_newdocs.py:
```diff
@@ -2480,68 +2634,16 @@ def luf(lamdaexpr, *args, **kwargs):
-    30
-    >>> 1*4 + 4*1 + 5*2 + 6*2
-    30
+    >>> path, desc = np.einsum_path('ij,ki,lj->lk', a, b, a, optimize='optimal')
```
I don't find this example particularly clarifying. In part, because `a` and `b` are defined above, for another example. In part, because it doesn't actually show any improvement in timing.

attack68 replied:
I confess I didn't spend too much time trying to contrive a better example. In my experience the process was discovered when I had a relatively complex case and it was the action of pre-computing the path I was interested in showing, not necessarily the calc variables. If I have time to think up a better use case I'll add one in, instead.

mhvk reviewed (commented):
Overall, nice clarification! I only find the example of `optimize` less than clear.

Also, maybe not for this PR, but having the same docstring in two different places seems like a bad idea...

attack68 committed:
DOC: minor requested corrections

attack68 commented:
I was trying to come up with a better example of optimize in iterative loops and I discovered this:

    import numpy as np
    a = np.arange(64).reshape(2,4,8)
    np.einsum('obk,ijk->ioj',a,a)
    array([[[ 1904,  5872,  9840, 13808],
            [ 5488, 17648, 29808, 41968]],
           [[17776, 21744, 25712, 29680],
            [54128, 66288, 78448, 90608]]])
    np.einsum('obk,ijk->ioj',a,a, optimize=True)
```
    Traceback (most recent call last):
  File "/anaconda3/envs/General/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2963, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-133-00a49d0e7c9d>", line 1, in <module>
    np.einsum('obk,ijk->ioj',a,a, optimize=True)
  File "/anaconda3/envs/General/lib/python3.6/site-packages/numpy/core/einsumfunc.py", line 1118, in einsum
    if any(left_dims[ind] != right_dims[ind] for ind in idx_rm):
  File "/anaconda3/envs/General/lib/python3.6/site-packages/numpy/core/einsumfunc.py", line 1118, in <genexpr>
    if any(left_dims[ind] != right_dims[ind] for ind in idx_rm):
KeyError: 'b'
```
Is this a known error on v 1.14.3?

charris commented:
Still a problem in master. We are waiting for some fixes from @dgasmith at https://github.com/dgasmith/opt_einsum. You might try that to see if it is fixed there.

attack68 committed:
DOC: suggested optimize argument Example cases

attack68 committed:
DOC: copy amendments to second docstring location in add_newdocs.py

mhvk commented:
@attack68 - yes, exactly the kind of example I was looking for. Thanks!

dgasmith commented on numpy/core/einsumfunc.py:
```diff
@@ -1051,13 +1137,26 @@ def einsum(*operands, **kwargs):
+    Chained array operations. For more complicated contractions, speed ups
+    might be achieved by repeatedly computing a 'greedy' path or pre-computing the
+    'optimal' path and repeatedly applying it, using an
+    `einsum_path` insertion (since version 1.12.0):
```
It might be good to note that the speedups usually become dramatically larger as the size of the tensors increase.

dgasmith commented:
Thanks for tackling this, it looks great overall. 

I can confirm `'obk,ijk->ioj'` is fixed in `opt_einsum`. We need to resync these codes soon.

```
>>> np.einsum('ijk,ilm,njm,nlk,abc->',a,a,a,a,a, optimize='greedy')
array(262144.)
```

Interesting, it depends on the last caller. `np.tensordot` always returns an array. While `dot/vdot/einsum` returns floats. I would argue that we would probably want to fix `tensordot` to be more consistent with other parts of the ecosystem. 

```
>>> np.tensordot(np.arange(5), np.arange(5), axes=[[0],[0]])
array(30)
```
Can you make an issue about these two bugs you found in `einsum` and ping me.

attack68 committed:
DOC: more descriptive text

attack68 committed:
DOC: required correction to specify a default argument.

attack68 commented:
I believe this has evolved to a state that now addresses all the above points, and dgasmith has committed fixes for the detected problems. 

Need I do more or just wait for it to process (this is my first numpy pull request so do not know the procedure)?

mattip commented:
@dgasmith could you give this a final review/ok?

dgasmith commented on numpy/add_newdocs.py:
```diff
@@ -2265,25 +2265,86 @@ def luf(lamdaexpr, *args, **kwargs):
+    dot(`a`, `b`).  If the first argument is complex the complex conjugate
+    of the first argument is used for the calculation of the dot product.
+
+    Note that `vdot` handles multidimensional arrays differently than `dot`:
```
It could be cleaner to cast this discussion in terms of inner products like `i,i->` and `ij,ij->`. I think the inner product language may be slightly more universal.

attack68 replied:
I agree, but actually I didn't amend this docstring. It was thought I should remove this sec but then advised to re-add, and when I did I inadvertently changed the order of functions, so this `add_newdoc` `vdot` now comes before `c_einsum` instead of after, the corresponding deletions are below, unchanged. I would prefer to just leave this as is. For clarity I have only actually altered the `einsum` docstring.

dgasmith replied:
Works for me, I think this ready for merge!

dgasmith reviewed (approved):
Overall LGTM, minor comment that might help. But I am fine without the addition.

Thanks for the doc improvement, this looks great!

mattip commented:
The documentation of `c_einsum` does not render in the html/pdf documentation, it is only available from `np.core.multiarray.c_einsum.__doc__`. I guess that is alright, since it is a 1:1 copy of the `einsum` documentation, except for the `optimization` keyword argument.

We do not render functions in the `multiarray` module directly, rather only as they appear in `numpy` itself, s part of an `__all__` directive. We could add a line somewhere in `doc/source/reference/routines.linalg.rst`, it would have to be `core.multiarray.c_einsum`, but I am disinclined since it would be the only such function from multiarray documented in this way.

attack68 commented:
There was a comment above about the problem of dual documentation on this. It seems I have fallen foul of the trap and didn't really understand its purpose but now I understand why its in two places.

I would recommend I do the following to the c_einsum docs:

* change the text `einsum` to `c_einsum`.
* remove the optimize argument in `c_einsum` docs.
* remove the example and related comments to the optimize argument in the `c_einsum` docstring.
* make a note in the `c_einsum` docstring that it has been copied from the other `einsum` docstring and augmented to reflect a lack of optimize argument.

I won't touch the `routines.linalg.rst` since I don't want to complicate matters.

Is this sensible or you prefer leave as is?

attack68 committed:
DOC: minor presentation fix for example display in einsumfunc.py

attack68 committed:
DOC: amend c_einsum documentation version, add note about dual documentation

attack68 commented on numpy/add_newdocs.py:
```diff
@@ -2265,25 +2265,86 @@ def luf(lamdaexpr, *args, **kwargs):
-    c_einsum(subscripts, *operands, out=None, dtype=None, order='K', casting='safe')
+    einsum(subscripts, *operands, out=None, dtype=None, order='K',
+           casting='safe', optimize=False)
 
```
take a look at commit c66f57 to see if you agree this a sensible way of handling this. Also I made the "c_einsum" correction.

mattip commented:
LGTM

mattip commented:
@charris do we usually squash a PR like this into one changeset before merging?

charris commented:
@mattip With 16 commits doing small things and many with similar commit messages correcting previous commits, I think squashing is appropriate. It's a judgement thing, what tends to push me to squash, apart from number of commits, would be bad commit messages,  lack of orthogonality in the commits, or intent to backport.

mattip commented:
@attack68, could you squash this to one commit?

charris commented:
@mattip you can squash the merge, it is one of the merge options, hit the down arrow.
